src/System.Net.Requests/src/System/Net/FtpWebRequest.cs

```diff
@@ -0,0 +1,2544 @@
+
+#if FEATURE_FTP_CACHE
+                        isRevalidatedOrRetried =!m_CacheDone &&
+                                                (CacheProtocol.ProtocolStatus == CacheValidationStatus.Continue || CacheProtocol.ProtocolStatus == CacheValidationStatus.RetryResponseFromServer);
+
+                        // This is for sync Upload commands that do not get chance hit GetResponse loop
+                        if (m_MethodInfo.IsUpload)
+                        {
+                            CheckCacheRetrieveOnResponse();
+                            CheckCacheUpdateOnResponse();
+                        }
+#endif
+                    }
+
+#if FEATURE_FTP_CACHE
+                    if (!isRevalidatedOrRetried)
+#endif
+                        stageMode = RequestStage.ReleaseConnection;
+                }
+            }
+            catch (Exception exception) {
+                SetException(exception);
+            }
+            finally {
+                FinishRequestStage(stageMode);
+                if (GlobalLog.IsEnabled)
+                {
+                    GlobalLog.Leave("FtpWebRequest#" + ValidationHelper.HashString(this) + "::SyncRequestCallback");
+                }
+                CheckError(); //will throw on error
+            }
+        }
+        //
+        // Only executed for Async requests
+        //
+        private void AsyncRequestCallback(object obj)
+        {
+            if (GlobalLog.IsEnabled)
+            {
+                GlobalLog.Enter("FtpWebRequest#" + ValidationHelper.HashString(this) + "::AsyncRequestCallback", "#" + ValidationHelper.HashString(obj));
+            }
+            RequestStage stageMode = RequestStage.CheckForError;
+
+            try {
+
+                FtpControlStream connection;
+                connection = obj as FtpControlStream;
+                FtpDataStream stream = obj as FtpDataStream;
+                Exception exception = obj as Exception;
+
+                bool completedRequest = (obj == null);
+
+                if (GlobalLog.IsEnabled)
+                {
+                    GlobalLog.Print("AsyncRequestCallback()  stream:" + ValidationHelper.HashString(stream) + " conn:" + ValidationHelper.HashString(connection) + " exp:" + ValidationHelper.HashString(exception) + " completedRequest:" + ValidationHelper.HashString(completedRequest));
+                }
+                while (true)
+                {
+                    if (exception != null)
+                    {
+                        if (AttemptedRecovery(exception))
+                        {
+                            connection = CreateConnection();
+                            if (connection == null)
+                                return;
+
+                            exception = null;
+                        }
+                        if (exception != null)
+                        {
+                            SetException(exception);
+                            break;
+              
```

Comment:
Nit: the & here would break the XML comment if we were actually compiling XML comments into docs